Issue with asar packaging (after 'prolonged' use, files start to ENOENT)

This issue occurs on both Electron v30.0 and v33.0 but only reproducible on Windows (x64). We have tested extensively on OSX and can't reproduce but Windows 32-bit and Linux are untested.

Essentially we have an application with lots of image resources packaged into the .asar archive. Our application will frequently load images from the archive - depending on state in our application, we completely unload assets from memory so we are often re-requesting them. We have used both XHR + file:// requests and the `fs` module directly and the result is the same.

After some time (ie continued / frequent requests of assets) - the files start to ENOENT as if they don't exist. Once the files start to ENOENT, they can never be read again unless the application is restarted. How long this takes to occur may vary but it is almost 100% reproducible for us.

If we disable asar packing and just bundle without the asar, the issue does not occur. 

Not clear if the issue lies inside Electron's `asar.coffee` or the asar implementation itself. It does not appear the asar archive becomes corrupt (as it works fine when you restart) but something is breaking the reading of files.

You basically see something like below, despite the fact that all the files obviously exist:
```
vendor.js:11 Error: ENOENT, src\resources\units\neutral_beastdarkharbinger.png not found in C:\Users\m\app\v1.0.18\resources\app.asar(…)
vendor.js:11 Error: ENOENT, src\resources\units\neutral_golembloodshard.png not found in C:\Users\m\app\v1.0.18\resources\app.asar(…)
vendor.js:11 Error: ENOENT, src\resources\units\neutral_mechaz0rcannon.png not found in C:\Users\m\app\v1.0.18\resources\app.asar(…)
vendor.js:11 Error: ENOENT, src\resources\units\neutral_chaoselemental.png not found in C:\Users\m\app\v1.0.18\resources\app.asar(…)
vendor.js:11 Error: ENOENT, src\resources\units\neutral_monsteroculus.png not found in C:\Users\m\app\v1.0.18\resources\app.asar(…)
vendor.js:11 Error: ENOENT, src\resources\units\neutral_golem_crossbones.png not found in C:\Users\m\app\v1.0.18\resources\app.asar(…)
vendor.js:11 Error: ENOENT, src\resources\units\neutral_golemstone.png not found in C:\Users\m\app\v1.0.18\resources\app.asar(…)
```

I can try to write up a standalone test case that reproduces this behavior if necessary.
Just again tested on 33.1 (noting the other bug fix in the asar code) but it still happens.

--- a/atom/common/asar/archive.cc
+++ b/atom/common/asar/archive.cc
@@ -115,6 +115,14 @@ bool FillFileInfoWithNode(Archive::FileInfo* info,
 Archive::Archive(const base::FilePath& path)
     : path_(path),
       file_(path_, base::File::FLAG_OPEN | base::File::FLAG_READ),
+#if defined(OS_WIN)
+      fd_(_open_osfhandle(
+              reinterpret_cast<intptr_t>(file_.GetPlatformFile()), 0)),
+#elif defined(OS_POSIX)
+      fd_(file_.GetPlatformFile()),
+#else
+      fd_(-1),
+#endif
       header_size_(0) {
 }
 
@@ -271,17 +279,7 @@ bool Archive::CopyFileOut(const base::FilePath& path, base::FilePath* out) {
 }
 
 int Archive::GetFD() const {
-  if (!file_.IsValid())
-    return -1;
-
-#if defined(OS_WIN)
-  return
-    _open_osfhandle(reinterpret_cast<intptr_t>(file_.GetPlatformFile()), 0);
-#elif defined(OS_POSIX)
-  return file_.GetPlatformFile();
-#else
-  return -1;
-#endif
+  return fd_;
 }
 
 }  // namespace asar

--- a/atom/common/asar/archive.h
+++ b/atom/common/asar/archive.h
@@ -69,6 +69,7 @@ class Archive {
  private:
   base::FilePath path_;
   base::File file_;
+  int fd_;
   uint32 header_size_;
   scoped_ptr<base::DictionaryValue> header_;
 

--- a/spec/api-browser-window-spec.coffee
+++ b/spec/api-browser-window-spec.coffee
@@ -294,16 +294,6 @@ describe 'browser-window module', ->
       w.show()
       w.minimize()
 
-  describe 'will-navigate event', ->
-    @timeout 10000
-    it 'emits when user starts a navigation', (done) ->
-      url = "file://#{fixtures}/pages/will-navigate.html"
-      w.webContents.on 'will-navigate', (event, u) ->
-        event.preventDefault()
-        assert.equal u, url
-        done()
-      w.loadUrl url
-
   xdescribe 'beginFrameSubscription method', ->
     it 'subscribes frame updates', (done) ->
       w.loadUrl "file://#{fixtures}/api/blank.html"

--- a/spec/asar-spec.coffee
+++ b/spec/asar-spec.coffee
@@ -8,6 +8,10 @@ describe 'asar package', ->
 
   describe 'node api', ->
     describe 'fs.readFileSync', ->
+      it 'does not leak fd', ->
+        for i in [1..10000]
+          fs.readFileSync(path.join(process.resourcesPath, 'atom.asar', 'renderer', 'api', 'lib', 'ipc.js'))
+
       it 'reads a normal file', ->
         file1 = path.join fixtures, 'asar', 'a.asar', 'file1'
         assert.equal fs.readFileSync(file1).toString().trim(), 'file1'

